### PR TITLE
Improve logs when connector unable to migrate.

### PR DESCRIPTION
**Pull request by Harkishen-Singh**
Signed-off-by: Harkishen-Singh <harkishensingh@hotmail.com>

Fixes: #444

### CODE REVIEW

**Review by blagojts — APPROVED** [review]

File: pkg/util/lock.go [hunk 1]
```diff
@@ -25,7 +25,7 @@ var (
 func GetSharedLease(ctx context.Context, conn *pgx.Conn, id int64) error {
 	gotten, err := runLockFunction(ctx, conn, "SELECT pg_try_advisory_lock_shared($1)", id)
 	if err != nil {
-		return err
+		return fmt.Errorf("unable to get schema lock. Please make sure that no Promscale connectors are running: %w", err)
```
Comment (cevian) [inline pkg/util/lock.go:28]:
```suggestion
		return fmt.Errorf("Unable to get shared schema lock. Please make sure that no operations requiring exclusive locking : %w", err)
```

File: pkg/pgmodel/migrate.go [hunk 2]
```diff
@@ -122,7 +122,7 @@ func CheckSchemaVersion(ctx context.Context, conn *pgx.Conn, versionInfo Version
 		return fmt.Errorf("failed to check schema version: %w", err)
 	}
 	if dbVersion.Compare(expectedVersion) != 0 {
-		return fmt.Errorf("db schema version is incorrect: expected %v, got %v", expectedVersion, dbVersion)
+		return fmt.Errorf("error while comparing schema version. This can cause issues while schema migration. Received schema version %v but expected %v. Please note that no Promscale connectors are running that can prevent from schema migration due to unavailability of schema lock", expectedVersion, dbVersion)
```
Comment (cevian) [inline pkg/pgmodel/migrate.go:125]: We should only print this message if we couldn't get the schema lock. That means we need to pass the variable that says whether or not the exclusive schema lock was obtained down to here

**Review by cevian — CHANGES_REQUESTED** [review]

File: pkg/pgmodel/migrate.go [hunk 3]
```diff
@@ -107,22 +107,24 @@ func Migrate(db *pgx.Conn, versionInfo VersionInfo, extOptions extension.Extensi
...
+	if err = CheckSchemaVersion(context.Background(), db, versionInfo, true); err != nil { // We pass hasSharedLease as true since it has already been verified at the start.
 		return err
 	}
-
 	return extension.CheckVersions(db, migrationFailedDueToLockError, extOptions)
 }
 
 // CheckSchemaVersion checks the DB schema version without checking the extension
-func CheckSchemaVersion(ctx context.Context, conn *pgx.Conn, versionInfo VersionInfo) error {
+func CheckSchemaVersion(ctx context.Context, conn *pgx.Conn, versionInfo VersionInfo, hasSharedLease bool) error {
```
Comment (cevian) [inline pkg/pgmodel/migrate.go:117]: I think there is confusion here what we care about is not the shared lease lock but the exclusive migration lock

File: pkg/pgmodel/migrate.go [hunk 4]
```diff
@@ -107,22 +107,24 @@ func Migrate(db *pgx.Conn, versionInfo VersionInfo, extOptions extension.Extensi
 // the extension, are set up correctly. This will set the ExtensionIsInstalled
 // flag and thus should only be called once, at initialization.
 func CheckDependencies(db *pgx.Conn, versionInfo VersionInfo, migrationFailedDueToLockError bool, extOptions extension.ExtensionMigrateOptions) (err error) {
-	if err = CheckSchemaVersion(context.Background(), db, versionInfo); err != nil {
+	if err = CheckSchemaVersion(context.Background(), db, versionInfo, true); err != nil { // We pass hasSharedLease as true since it has already been verified at the start.
```
Comment (cevian) [inline pkg/pgmodel/migrate.go:110]: What we should be passing down is `migrationFailedDueToLockError` not anything about the shared lease

File: pkg/pgmodel/migrate.go [hunk 5]
```diff
@@ -107,22 +107,24 @@ func Migrate(db *pgx.Conn, versionInfo VersionInfo, extOptions extension.Extensi
...
+func CheckSchemaVersion(ctx context.Context, conn *pgx.Conn, versionInfo VersionInfo, hasSharedLease bool) error {
 	expectedVersion := semver.MustParse(versionInfo.Version)
 	dbVersion, err := getSchemaVersionOnConnection(ctx, conn)
 	if err != nil {
 		return fmt.Errorf("failed to check schema version: %w", err)
 	}
 	if dbVersion.Compare(expectedVersion) != 0 {
-		return fmt.Errorf("db schema version is incorrect: expected %v, got %v", expectedVersion, dbVersion)
+		if !hasSharedLease {
+			return fmt.Errorf("Error while comparing schema version. Failed to acquire shared schema lock. Please ensure that no Promscale connectors are running that can prevent from schema migration due to unavailability of shared schema lock. Received schema version %v but expected %v", dbVersion, expectedVersion)
```
Comment (cevian) [inline pkg/pgmodel/migrate.go:125]:
```suggestion
			return fmt.Errorf("Failed to acquire the migration lock to upgrade the schema version and unable to run with the old version. Please ensure that no other Promscale connectors with the old schema version are running. Received schema version %v but expected %v", dbVersion, expectedVersion)
```

File: pkg/pgmodel/migrate.go [hunk 6]
```diff
@@ -107,22 +107,24 @@ func Migrate(db *pgx.Conn, versionInfo VersionInfo, extOptions extension.Extensi
...
-func CheckSchemaVersion(ctx context.Context, conn *pgx.Conn, versionInfo VersionInfo) error {
+func CheckSchemaVersion(ctx context.Context, conn *pgx.Conn, versionInfo VersionInfo, hasSharedLease bool) error {
 	expectedVersion := semver.MustParse(versionInfo.Version)
 	dbVersion, err := getSchemaVersionOnConnection(ctx, conn)
 	if err != nil {
 		return fmt.Errorf("failed to check schema version: %w", err)
 	}
 	if dbVersion.Compare(expectedVersion) != 0 {
-		return fmt.Errorf("db schema version is incorrect: expected %v, got %v", expectedVersion, dbVersion)
+		if !hasSharedLease {
```
Comment (cevian) [inline pkg/pgmodel/migrate.go:124]:
```suggestion
		if  dbVersion.Compare(expectedVersion)  < 0 && !hasSharedLease {
```

File: pkg/runner/runner.go [hunk 7]
```diff
@@ -406,9 +406,14 @@ func compileAnchoredRegexString(s string) (*regexp.Regexp, error) {
...
+		hasSharedLease = true
+		err            = util.GetSharedLease(ctx, conn, schemaLockId)
+	)
+	if err == util.SharedLeaseFailure {
+		hasSharedLease = false
+	} else if err != nil {
+		return fmt.Errorf("get schema lease: %w", err)
 	}
-	return pgmodel.CheckSchemaVersion(ctx, conn, appVersion)
+	return pgmodel.CheckSchemaVersion(ctx, conn, appVersion, hasSharedLease)
```
Comment (cevian) [inline pkg/runner/runner.go:418]: yeah this is not the path we are worried about. The shared lease doesn't matter much

**Review by cevian — CHANGES_REQUESTED** [review]

File: pkg/util/lock.go [hunk 8]
```diff
@@ -18,14 +18,12 @@ import (
...
-var (
-	SharedLeaseFailure = fmt.Errorf("failed to acquire shared lease")
-)
+var SharedLeaseFailure = fmt.Errorf("failed to acquire shared lease")
 
 func GetSharedLease(ctx context.Context, conn *pgx.Conn, id int64) error {
 	gotten, err := runLockFunction(ctx, conn, "SELECT pg_try_advisory_lock_shared($1)", id)
 	if err != nil {
-		return err
+		return fmt.Errorf("Unable to get shared schema lock. Please make sure that no operations requiring exclusive locking: %w", err)
```
Comment (cevian) [inline pkg/util/lock.go:26]:
```suggestion
		return fmt.Errorf("Unable to get shared schema lock. Please make sure that no operations requiring exclusive locking are running: %w", err)
```

File: pkg/pgmodel/migrate.go [hunk 9]
```diff
@@ -107,22 +107,24 @@ func Migrate(db *pgx.Conn, versionInfo VersionInfo, extOptions extension.Extensi
...
-func CheckSchemaVersion(ctx context.Context, conn *pgx.Conn, versionInfo VersionInfo) error {
+func CheckSchemaVersion(ctx context.Context, conn *pgx.Conn, versionInfo VersionInfo, migrationFailedDueToLockError bool) error {
 	expectedVersion := semver.MustParse(versionInfo.Version)
 	dbVersion, err := getSchemaVersionOnConnection(ctx, conn)
 	if err != nil {
 		return fmt.Errorf("failed to check schema version: %w", err)
 	}
-	if dbVersion.Compare(expectedVersion) != 0 {
-		return fmt.Errorf("db schema version is incorrect: expected %v, got %v", expectedVersion, dbVersion)
+	if versionState := dbVersion.Compare(expectedVersion); versionState != 0 {
```
Comment (cevian) [inline pkg/pgmodel/migrate.go:123]:
```suggestion
	if versionCompare := dbVersion.Compare(expectedVersion); versionCompare != 0 {
```

**Review by cevian — APPROVED** [review]
Minor changes asked, but good to merge otherwise. Thanks!

**Pull request merged**